Adding Orb command

ochaloup opened the pull request:
I'm proposing low level commands (online/offline) for orb subsystem.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java:
```diff
@@ -0,0 +1,82 @@
+ */
+class Attribute<T> {
+    private T value;
+    private boolean isToUndefine = false;
```
Should be easy to make the class `final` and these two fields `final` as well.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/AuthValues.java:
```diff
@@ -0,0 +1,18 @@
+    CLIENT_AUTH("ClientAuth"),
+    MUTUAL_AUTH("MutualAuth");
+
+    public final String value;
```
Not `public`.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java:
```diff
@@ -0,0 +1,606 @@
+ *
+ * <ul>
+ *   <li>jacorb for EAP 6.x (<= WFLY 8)
+ *   <li>openjdk iiop for EAP 7.x (>= WFLY 9)
```
I'd prefer stating the upstream relationship first. If you want to add EAP relationship, that's fine, but _that_ belongs to the parens. Like:

```java
<li>JacORB for AS7 and WildFly <= 8 (EAP 6)</li>
<li>OpenJDK ORB for WildFly >= 9 (EAP 7)</li>
```

Personally, I don't feel the EAP bits are needed, but feel free to keep them.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java:
```diff
@@ -0,0 +1,82 @@
+     * Defines there is a value for this attribute.<br/>
+     * It does <b>not</b> consider if this attribute should or should not be undefined.
+     */
+    public boolean havingValue() {
```
`hasValue`

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java:
```diff
@@ -0,0 +1,82 @@
+     * Defines there is no value for this attribute.<br/>
+     * It does <b>not</b> consider if this attribute should or should not be undefined.
+     */
+    public boolean notHavingValue() {
```
`hasNoValue` or `doesntHaveValue`

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java:
```diff
@@ -0,0 +1,606 @@
+
+    static enum OrbType {
+        // jacorb in EAP 6.4.x series and lower
+        EAP_JACORB,
```
`AS7_JACORB`

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java:
```diff
@@ -0,0 +1,606 @@
+    static enum OrbType {
+        // jacorb in EAP 6.4.x series and lower
+        EAP_JACORB,
+        // jacorb in WFLY 8 and 9
```
So what is it? In `ChangeOrb` javadoc, you say that JacORB is <= WildFly 8 and OpenJDK ORB is WildFly >= 9.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java:
```diff
@@ -0,0 +1,606 @@
+        }
+
+
+
```
Why so many empty lines? :-)

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOffValues.java:
```diff
@@ -0,0 +1,18 @@
+    OFF(false, "off");
+
+    public final Boolean openjdk;
+    public final String jacorb;
```
Not `public`.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOffValues.java:
```diff
@@ -0,0 +1,18 @@
+package org.wildfly.extras.creaper.commands.orb;
+
+public enum OnOffValues {
```
`OnOff`

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/SecurityValues.java:
```diff
@@ -0,0 +1,15 @@
+    NONE("none", "off");
+
+    public final String openjdk;
+    public final String jacorb;
```
Not `public`.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/SupportedValues.java:
```diff
@@ -0,0 +1,17 @@
+    SUPPORTED("supported"),
+    REQUIRED("required");
+
+    public final String value;
```
Not `public`.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/TransactionValues.java:
```diff
@@ -0,0 +1,15 @@
+    SPEC("spec", "spec");
+
+    public final String openjdk;
+    public final String jacorb;
```
Not `public`.

Ladicek commented:
I wonder why the test failed, but it shouldn't be caused by the changes in this PR. I'll look into it.

Ladicek commented:
Could you please rebase on master? I've fixed the failing test.

ochaloup commented:
I'm sorry for overhelming the ci server with wrong pushs. I changed the code at way you requested. Hopefully it's ok.

Regarding you remark on using `Attribute` class. I though that this is the right way how to create a command - having some low level that could be used for later addition of high level ones. I was absorbed with this idea that I spent quite significant effort to implement that way.

But with your note when looking back I have to say I'm now unsure if it was right path then. I can image that it would be nice to have all commands implemented in similar way but I understand it's first probably impossible because of time limitation and second I could be alone with this objective.

I fully understand that this way the command is different from other ones.
If you be happier if I split command to two `Change` and `Undefine` for not having the `Attribute` functionality involved I can do so.
On the other hand the public api does not deal with the trouble of some strange `Attribute` class being seen as you demanded.
What is your stand point on this?

ochaloup commented:
ops, I'm reviewing old version of your notes. I need to need a bit more changes. Somehow overlooked that github is not in sync.

Ladicek commented:
No, no need to change the architecture. Noone cares about configuring ORB anyway :-), so the minor difference doesn't really matter.

ochaloup commented:
...now I'm not sure why tests failed. I run on my machine and it passes.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/AuthValues.java:
```diff
@@ -0,0 +1,22 @@
+
+    public String getValue() {
+        return value;
+    }
```
Why do you need this `public`? Either this method should be package-private or the `value` field should be package-private. I guess this also applies to other enums in this PR.

ochaloup replied:
ah, sure. I had to be unconscious when fixing the issue last time.

...don't worry master, I won't disappoint you again :]

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java:
```diff
@@ -0,0 +1,606 @@
+    }
+
+    enum OrbType {
+        // jacorb in AS7 and WFLY 7
```
There's no WFLY 7.

ochaloup replied:
after so many years I should find out they were lying me? impossible!

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOff.java:
```diff
@@ -0,0 +1,26 @@
+    public String getJacorb() {
+        return jacorb;
+    }
+}
```
All `Boolean`s in this class could apparently be `boolean`s?

Ladicek commented on testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/orb/ChangeOrbOnlineTest.java:
```diff
@@ -0,0 +1,232 @@
+
+@Category(ManualTests.class)
+@RunWith(Arquillian.class)
+@RunAsClient
```
`@RunAsClient` is useless, please remove. The warnings from Arquillian that there's no running deployment are a bug, see [ARQ-2012](https://issues.jboss.org/browse/ARQ-2012).

ochaloup replied:
Personally I like usage of `@RunAsClient` as denotes if the test was written with notion of running on client or not. But of course removing by your request :smile: 
Thanks for pointing to the issue.

Ladicek commented:
Yea, I don't know why the test failed either. I'll try to figure it out.

ochaloup commented:
ok, hopefully all requests fixed.

Unfortunately ci fails again. It passes locally for me. No much idea how to get some more info from travis ci...

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java:
```diff
@@ -0,0 +1,82 @@
+        return new Attribute<T>(null, true);
+    }
+
+    public static <O> Attribute<O> undefine(O object) {
```
Hmm, I wonder what is this good for? The `object` is never used in this method. And the method seems to be unused as well.

If you wanted a type-safe method and the `<T> Attribute<T> undefine()` variant isn't enough (which might happen in Java pre 8), idiomatic solution would be `<T> Attribute<T> undefine(Class<? extends T> clazz)`.

ochaloup replied:
it isn't used, I've removed it. It was left in code from pre-pull request dates where it was used in command builder.

my lack of knowledge. I was thinking about `Class<? extends T> clazz` but I got some trouble to do so which I can't remember. And I didn't know that is considered to be an idiomatic solution.
Btw. why this does not happen for Java 8 (for not stealing your time, some link would be enough :)?

btw. a longer time I'm wondering why `Arrays#toArray(T[] a)` method (https://docs.oracle.com/javase/7/docs/api/java/util/Collection.html#toArray(T[])) is not defined in similar way as you mentioned?

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java:
```diff
@@ -0,0 +1,606 @@
+ *   <li>OpenJDK ORB for WildFly >= 9</li>
+ * </ul>
+ */
+public class ChangeOrb implements OfflineCommand, OnlineCommand {
```
`final`?

ochaloup replied:
makes sence

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java:
```diff
@@ -0,0 +1,606 @@
+
+    private static final class AttributeWriter {
+        private Address address;
+        private Batch batch;
```
Both should be `final`.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOff.java:
```diff
@@ -0,0 +1,18 @@
+    final boolean openjdk;
+    final String jacorb;
+
+    OnOff(Boolean openjdk, String jacorb) {
```
`boolean`?

ochaloup replied:
agree, I was too rash

ochaloup commented:
final adjustments

...final? LOL!

Ladicek commented:
Tests seem to be failing on WildFly 8 and 9, please run `mvn clean verify -Pwildfly8` (resp. `-Pwildfly9`). Otherwise it's looking fine.

Oh, maybe you could fix the typo in the commit message -- it's not `codearc`, it's `CodeNarc` :-)

ochaloup commented:
Commit message fixed.

Fixed the trouble with wfly8 and wfly9 - simplifier the options to have only jacorb or iiop. I expect now it should be working fine for all versions.

Ladicek commented:
I tried to improve the `ShutdownTest` a bit, could you please rebase?

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOff.java:
```diff
@@ -0,0 +1,18 @@
+package org.wildfly.extras.creaper.commands.orb;
+
+public enum OnOff {
```
This doesn't seem to be exposed anywhere, so it shouldn't be `public`. Other enums seem to be used, so it's just this one.

ochaloup commented:
Hiding OnOff and the rebase done.

Ladicek commented:
Merged, finally :-) Thanks for the patience ;-)

ochaloup commented:
My thanks to you.

 ...for patience ;) and for good advices!
